### PR TITLE
Change recommendations (WIP)

### PR DESCRIPTION
**Pull request by thoessl**

### CODE REVIEW

File: openslides/motions/models.py [hunk 1]
```diff
@@ -697,6 +698,57 @@ def get_root_rest_element(self):
         return self.motion
 
 
+class MotionChangeRecommendation(RESTModelMixin, models.Model):
+    """
+    A MotionChangeRecommendation object saves change recommendations for a specific MotionVersion
+    """
+
+    access_permissions = MotionChangeRecommendationAccessPermissions()
+
```
Comment (normanjaeckel) [inline openslides/motions/models.py]: Please add a custom manager with the get_full_queryset method, see most other models. This method is to improve performance during cache filling.

File: openslides/motions/models.py [hunk 2]
```diff
@@ -697,6 +698,57 @@ def get_root_rest_element(self):
...
+    motion_version = models.ForeignKey(
+        MotionVersion,
+        on_delete=models.CASCADE,
+        related_name='change_recommendations')
+    """The motion version to which the change recommendation belongs."""
+
+    type = models.PositiveIntegerField()
+    """ Insertion (1), Deletion (2) or Replacement (0) """
```
Comment (normanjaeckel) [inline openslides/motions/models.py]: Docstring should be consistent. Sometimes they start with whitespace, sometimes not.

Comment (normanjaeckel) [inline openslides/motions/models.py]: I wonder what "insertation" together with line_from==23 and line_to==42 should mean. Do we get an inconsistent database here?

(I always tried to argue that we have only one type at all: replacement. Every change recommendation is a replacement. Either I replace "foo bar" by "foo baz bar" with looks like an insertation or I replace "foo bar" by "foo" which looks like a deletion or I replace "foo" by "bar" which looks like a replacement. From the server's point of view we do not need to save the redundant type information.)

File: openslides/motions/models.py [hunk 3]
```diff
@@ -697,6 +698,57 @@ def get_root_rest_element(self):
...
+    """The replacement for the section of the original text specified by version, line_from and line_to """
+
+    author = models.ForeignKey(
+        settings.AUTH_USER_MODEL,
+        on_delete=models.SET_NULL,
+        null=True)
+    """A user object, who created this change recommendation. Optional."""
+
+    creation_time = models.DateTimeField(auto_now=True)
+    """Time when the version was saved."""
```
Comment (normanjaeckel) [inline openslides/motions/models.py]: It is not the version that is saved but the change recommendation, isn't it?

File: openslides/motions/models.py [hunk 4]
```diff
@@ -697,6 +698,57 @@ def get_root_rest_element(self):
...
+    class Meta:
+        default_permissions = ()
+
+    def __str__(self):
+        """Return a string, representing this object."""
+        return "Recommendation for Version %s, line %s - %s" % (self.motion_version_id, self.line_from, self.line_to)
+
+    def get_root_rest_element(self):
+        """
+        Returns the motion to this instance which is the root REST element.
```
Comment (normanjaeckel) [inline openslides/motions/models.py]: Docstring is wrong.

File: openslides/motions/static/js/motions/base.js [hunk 5]
```diff
@@ -187,7 +190,128 @@ angular.module('OpenSlidesApp.motions', [
 
                     return lineNumberingService.insertLineNumbers(html, lineLength, highlight, callback);
                 },
-                setTextStrippingLineBreaks: function (versionId, text) {
+
+                /* Not used currently
```
Comment (normanjaeckel) [inline openslides/motions/static/js/motions/base.js]: ?

File: openslides/motions/static/js/motions/base.js [hunk 6]
```diff
@@ -187,7 +190,128 @@ angular.module('OpenSlidesApp.motions', [
...
+                    return this.getTextWithChangeRecommendations(versionId, function(status) {
+                        return (status == 1);
+                    });
+                },
+                getTextWithoutRejectedChangeRecommendations: function (versionId) {
+                    return this.getTextWithChangeRecommendations(versionId, function(status) {
+                        return (status != 2);
+                    });
+                },
+
```
Comment (normanjaeckel) [inline openslides/motions/static/js/motions/base.js]: Remove at least one blank line.

File: openslides/motions/static/js/motions/base.js [hunk 7]
```diff
@@ -201,6 +325,24 @@ angular.module('OpenSlidesApp.motions', [
                 getSearchResultSubtitle: function () {
                     return "Motion";
                 },
+                /**
```
Comment (normanjaeckel) [inline openslides/motions/static/js/motions/base.js]: Here we have the docstring just before the function which is ok for me. But some lines later the docstring is inside. It should be consistent. So you can move the docstring inside the function or open an issue to cleanup all other functions.

File: openslides/motions/static/js/motions/site.js [hunk 8]
```diff
@@ -786,15 +897,19 @@ angular.module('OpenSlidesApp.motions.site', [
...
+             PollContentProvider, PdfMakeConverter, PdfMakeDocumentProvider, MotionInlineEditing, gettextCatalog,
+             Projector, HTMLValidizer, ProjectionDefault) {
         Motion.bindOne(motion.id, $scope, 'motion');
         Category.bindAll({}, $scope, 'categories');
         Mediafile.bindAll({}, $scope, 'mediafiles');
         Tag.bindAll({}, $scope, 'tags');
         User.bindAll({}, $scope, 'users');
         Workflow.bindAll({}, $scope, 'workflows');
+        MotionChangeRecommendation.bindAll({}, $scope, 'change_recommendations');
+        MotionChangeRecommendation.findAll();
```
Comment (normanjaeckel) [inline openslides/motions/static/js/motions/site.js]: Ah this should be done not here but in the resolver I think.

File: openslides/motions/static/templates/motions/motion-detail.html [hunk 9]
```diff
@@ -258,102 +258,62 @@ <h3 ng-if="motion.polls.length > 0" translate>Voting result</h3>
...
-        </label>
-        <label class="btn btn-sm btn-default" ng-class="{active: (lineNumberMode == 'outside')}"
-               ng-click="lineNumberMode = 'outside'">
-          <input type="radio" name="lineNumberMode" value="outside" ng-model="lineNumberMode"
-                 ng-checked="lineNumberMode == 'outside'">
-          <translate>outside</translate>
-        </label>
-      </span>
+    <!-- Motion toolbar -->
+    <ng-include src="'static/templates/motions/motion-detail-toolbar.html'"></ng-include>
```
Comment (normanjaeckel) [inline openslides/motions/static/templates/motions/motion-detail.html]: Proposal: Move all parts of motion detail to a subdirectory static/templates/motions/motion-detail/toolbar.html. So also the templates directory is much more cleaner.

File: openslides/motions/views.py [hunk 10]
```diff
@@ -341,6 +343,29 @@ def check_view_permissions(self):
...
+    access_permissions = MotionChangeRecommendationAccessPermissions()
+    queryset = MotionChangeRecommendation.objects.all()
+
+    def check_view_permissions(self):
+        """
+        Returns True if the user has required permissions. Obviously.
+        """
+        if self.action in ('list', 'retrieve'):
+            result = self.get_access_permissions().check_permissions(self.request.user)
+        elif self.action in ('create', 'destroy', 'update'):
```
Comment (normanjaeckel) [inline openslides/motions/views.py]: 'metadata' is missing. You need an extra else clause, see other viewsets.

File: openslides/motions/views.py [hunk 11]
```diff
@@ -341,6 +343,29 @@ def check_view_permissions(self):
...
+
+    There are the following views: metadata, list, retrieve, create,
+    partial_update, update and destroy.
+    """
+    access_permissions = MotionChangeRecommendationAccessPermissions()
+    queryset = MotionChangeRecommendation.objects.all()
+
+    def check_view_permissions(self):
+        """
+        Returns True if the user has required permissions. Obviously.
```
Comment (normanjaeckel) [inline openslides/motions/views.py]: :-)

File: openslides/motions/views.py (hunk 10 above)
Comment (normanjaeckel) [inline openslides/motions/views.py]: Also partial_update is missing.

File: openslides/motions/static/templates/motions/motion-detail.html (hunk 9 above)
Comment (emanuelschuetze) [inline openslides/motions/static/templates/motions/motion-detail.html]: No (as discussed with @normanjaeckel). Please irgnore this remark.

File: openslides/motions/models.py (hunk 1 above)
Comment (thoessl) [inline openslides/motions/models.py]: Should be resolved by https://github.com/OpenSlides/OpenSlides/pull/2501/commits/25a208bab2dbf0ac44dc8c451d3eb28972ce446d

File: openslides/motions/static/js/motions/site.js [hunk 12]
```diff
@@ -938,6 +1008,8 @@ angular.module('OpenSlidesApp.motions.site', [
         $scope.showVersion = function (version) {
             $scope.version = version.id;
             $scope.inlineEditing.setVersion(motion, version.id);
+            $scope.createChangeRecommendation.setVersion(motion, version.id);
+            console.log("show version");
```
Comment (normanjaeckel) [inline openslides/motions/static/js/motions/site.js]: Remove this line.

File: openslides/motions/static/templates/motions/motion-detail/change-summary.html [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+                <translate ng-if="change.status == 0">Suggested</translate>
+                <translate ng-if="change.status == 1">Accepted</translate>
+                <translate ng-if="change.status == 2">Rejected</translate>
+            </span>
+        </li>
+    </ul>
+    <div ng-if="change_recommendations.length == 0" class="no-changes">
+        <translate>No change recommendations yet</translate>
+    </div>
+</section>
```
Comment (normanjaeckel) [inline openslides/motions/static/templates/motions/motion-detail/change-summary.html]: Please let the file end with a newline.

File: openslides/motions/static/templates/motions/motion-detail/view-diff.html [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+
+                <div class="motion-text motion-text-diff line-numbers-{{ lineNumberMode }}"
+                     ng-bind-html="change.format(motion, version) | trusted"></div>
+            </div>
+        </div>
+
+        <div class="motion-text original-text line-numbers-{{ lineNumberMode }}"
+             ng-bind-html="motion.getTextRemainderAfterLastChangeRecommendation(version, changes) | trusted"></div>
+    </div>
+</div>
```
Comment (normanjaeckel) [inline openslides/motions/static/templates/motions/motion-detail/view-diff.html]: Please let the file end with a newline.

File: openslides/motions/static/templates/motions/motion-detail/view-original.html [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+            <!-- delete change recommendation -->
+            <button class="btn btn-default btn-xs"
+                    ng-bootbox-confirm="{{ 'Are you sure you want to delete this change recommendation?' | translate }}"
+                    ng-bootbox-confirm-action="viewChangeRecommendations.delete(change.id)"
+                    title="{{ 'Delete' | translate }}">
+                <i class="fa fa-times"></i>
+            </button>
+        </div>
+    </li>
+</ul>
```
Comment (normanjaeckel) [inline openslides/motions/static/templates/motions/motion-detail/view-original.html]: Please let the file end with a newline.

**Comment by emanuelschuetze** [comment]
Ok, successfully tested. Looks good now. 
@thoessl Please rebase to one commit today. Then I'll mege it. Thanks again!

**Comment by emanuelschuetze** [comment]
@CatoTH There is still a small conflict which I already resolved in my local branch. But I'm not allowed to push it directly into your branch. Please set this allowed-flag or rebase & fix it yourself.

**Pull request merged**